src/pstl_offload.cpp

```diff
@@ -61,7 +98,25 @@ __internal_free(void* __user_ptr)
         }
         else
         {
-            __get_original_free()(__user_ptr);
+            if (__dlsym_called)
+            {
+                // Delay releasing till exit of dlsym. We do not overload malloc globally,
+                // so can use it safely. Do not use new to able to use free() during
+                // __delayed_free_list releasing.
+                void* __buf = malloc(sizeof(__delayed_free_list));
+                if (!__buf)
+                {
+                    throw std::bad_alloc();
+                }
+                __delayed_free_list* __h = new(__buf) __delayed_free_list(__delayed_free, __user_ptr);
+                __delayed_free = __h;
+            }
+            else
+            {
+                static __free_func_type __orig_free = __get_original_free(__user_ptr);
+
+                __orig_free(__user_ptr);
```

Comment:
May be it would be better to rename `__get_original_free` to something like `__free_original(__user_ptr)` and move `__user_ptr` releasing inside of the function. I guess the logic would be more clear. What do you think?

Reply:
Symmetry with rest of `__get_original*` functions is a good thing, so I tend to agree.